File: spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/DataFrameTools.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iotdb.spark.db

import org.apache.iotdb.session.Session
import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType
import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.{BOOLEAN, DOUBLE, FLOAT, INT32, INT64, TEXT}
import org.apache.spark.sql.{DataFrame, Row}

import java.util
import java.lang

object DataFrameTools {
  def insertDataFrame(options: IoTDBOptions, dataframe: DataFrame): Unit = {
    val filteredColumns = Array[String]("Time", "Device")
    val sensorTypes = dataframe.dtypes.filter(x => !filteredColumns.contains(x._1))

    dataframe
      .repartition(options.numPartition.toInt)
      .sortWithinPartitions(dataframe.col("Device"))
      .foreachPartition { (partition: Iterator[Row]) =>
        val hostPort = options.url.split("//")(1).replace("/", "").split(":")
        val session = new Session(
          hostPort(0),
          hostPort(1).toInt,
          options.user,
          options.password
        )
        session.open()

        var device: lang.String = ""
        val times = new util.ArrayList[lang.Long]()
        val measurementsList = new util.ArrayList[util.List[lang.String]]()
        val typesList = new util.ArrayList[util.List[TSDataType]]()
        val valuesList = new util.ArrayList[util.List[Object]]()

        val batchSize = 1000
        var currentSize = 0

        partition.foreach { record =>
          if ("".equals(device)) device = record.get(1).toString
          else if (!device.equals(record.get(1).toString)) {
            insertAndEmptyDataSet(session, device, times, measurementsList, typesList, valuesList)
            device = record.get(1).toString
            currentSize = 0
          }
          val measurements = new util.ArrayList[lang.String]()
          val types = new util.ArrayList[TSDataType]()
          val values = new util.ArrayList[Object]()
          for (i <- 2 until record.length if !(record.get(i) == null)) {
            val value = typeTrans(record.get(i).toString, getType(sensorTypes(i - 2)._2))

            values.add(value)
            measurements.add(sensorTypes(i - 2)._1)
            types.add(getType(sensorTypes(i - 2)._2))
          }
          if (!values.isEmpty) {
            times.add(record.get(0).asInstanceOf[Long])
            measurementsList.add(measurements)
            typesList.add(types)
            valuesList.add(values)
            currentSize += 1
          }
          if (currentSize >= batchSize) {
            insertAndEmptyDataSet(session, device, times, measurementsList, typesList, valuesList)
            currentSize = 0
          }
        }

        insertAndEmptyDataSet(session, device, times, measurementsList, typesList, valuesList)
        session.close()
      }

  }

  def typeTrans(value: lang.String, dataType: TSDataType): Object = {
    dataType match {
      case TSDataType.TEXT => value
      case TSDataType.BOOLEAN => lang.Boolean.valueOf(value)
      case TSDataType.INT32 => lang.Integer.valueOf(value)
      case TSDataType.INT64 => lang.Long.valueOf(value)
      case TSDataType.FLOAT => lang.Float.valueOf(value)
      case TSDataType.DOUBLE => lang.Double.valueOf(value)
      case _ => null
    }
  }

  def getType(typeStr: lang.String): TSDataType = {
    typeStr match {
      case "StringType" => TEXT
      case "BooleanType" => BOOLEAN
      case "IntegerType" => INT32
      case "LongType" => INT64
      case "FloatType" => FLOAT
      case "DoubleType" => DOUBLE
      case _ => null
    }
  }

  def insertAndEmptyDataSet(session: Session,
                            device: lang.String,
                            times: util.ArrayList[lang.Long],
                            measurementsList: util.ArrayList[util.List[lang.String]],
                            typesList: util.ArrayList[util.List[TSDataType]],
                            valuesList: util.ArrayList[util.List[Object]]): Unit = {
    if (!valuesList.isEmpty) {
      session.insertRecordsOfOneDevice(device, times, measurementsList, typesList, valuesList)
      times.clear()
      measurementsList.clear()
      typesList.clear()
      valuesList.clear()
    }
  }
}

File: spark-iotdb-connector/src/main/scala/org/apache/iotdb/spark/db/tools/DataFrameTools.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iotdb.spark.db.tools;

import org.apache.iotdb.rpc.IoTDBConnectionException;
import org.apache.iotdb.rpc.StatementExecutionException;
import org.apache.iotdb.session.Session;
import org.apache.iotdb.spark.db.IoTDBOptions;
import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.thrift.annotation.Nullable;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.DOUBLE;
import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.BOOLEAN;
import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.FLOAT;
import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.INT32;
import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.INT64;
import static org.apache.iotdb.tsfile.file.metadata.enums.TSDataType.TEXT;

public class DataFrameTools {
  /**
   * * insert a narrow dataframe into IoTDB
   *
   * @param options the options to create a IoTDB session
   * @param dataFrame a dataframe of narrow table
   * @return
   */
  public static void insertDataFrame(IoTDBOptions options, Dataset<Row> dataFrame) {
    List<Tuple2<String, String>> sensorTypes = new ArrayList<>(Arrays.asList(dataFrame.dtypes()));
    sensorTypes.remove(0);
    sensorTypes.remove(0);

    List<Row> devices = dataFrame.select("device_name").distinct().collectAsList();

    Dataset<Row> repartition = dataFrame.repartition(dataFrame.col("device_name"));

    for (Row row : devices) {
      String device = row.get(0).toString();
      repartition
          .where(String.format("device_name == '%s'", device))
          .foreachPartition(
              partition -> {
                String[] hostPort = options.url().split("//")[1].replace("/", "").split(":");
                Session session =
                    new Session.Builder()
                        .host(hostPort[0])
                        .port(Integer.valueOf(hostPort[1]))
                        .username(String.valueOf(options.user()))
                        .password(String.valueOf(options.password()))
                        .build();
                session.open();
                partition.forEachRemaining(
                    record -> {
                      ArrayList<String> measurements = new ArrayList<>();
                      ArrayList<TSDataType> types = new ArrayList<>();
                      ArrayList<Object> values = new ArrayList<>();
                      for (int i = 2; i < record.length(); i++) {
                        Object value = record.get(i);
                        if (value == null) {
                          continue;
                        }
                        value =
                            typeTrans(record.get(i).toString(), getType(sensorTypes.get(i - 2)._2));

                        values.add(value);
                        measurements.add(sensorTypes.get(i - 2)._1);
                        types.add(getType(sensorTypes.get(i - 2)._2));
                      }
                      try {
                        session.insertRecord(
                            record.get(1).toString(),
                            (Long) record.get(0),
                            measurements,
                            types,
                            values);
                      } catch (IoTDBConnectionException e) {
                        e.printStackTrace();
                      } catch (StatementExecutionException e) {
                        e.printStackTrace();
                      }
                    });
                session.close();
              });
    }
  }
  /**
   * @param value
   * @param type
   * @return
   */
  private static Object typeTrans(String value, TSDataType type) {
    try {
      switch (type) {
        case TEXT:
          return value;
        case BOOLEAN:
          return Boolean.valueOf(value);
        case INT32:
          return Integer.valueOf(value);
        case INT64:
          return Long.valueOf(value);
        case FLOAT:
          return Float.valueOf(value);
        case DOUBLE:
          return Double.valueOf(value);
        default:
          return null;
      }
    } catch (NumberFormatException e) {
      return null;
    }
  }

  /**
   * return the TSDataType
   *
   * @param typeStr
   * @return
   */
  private static TSDataType getType(String typeStr) {
    switch (typeStr) {
      case "StringType":
        return TEXT;
      case "BooleanType":
        return BOOLEAN;
      case "IntegerType":
        return INT32;
      case "LongType":
        return INT64;
      case "FloatType":
        return FLOAT;
      case "DoubleType":
        return DOUBLE;
      default:
        return null;
    }
  }
}
